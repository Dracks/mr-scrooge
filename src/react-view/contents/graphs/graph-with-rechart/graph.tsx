import { Box, Button, Heading } from 'grommet';
import { Edit, FormDown, FormUp, Trash } from 'grommet-icons';
import React from 'react';
import { useAsyncCallback } from 'react-async-hook';
import { useNavigate } from 'react-router';

import { useApiClient } from '../../../api/client';
import { ApiUUID, Graph, MoveDirection } from '../../../api/models';
import { useLogger } from '../../../utils/logger/logger.context';
import { catchAndLog } from '../../../utils/promises';
import { ConfirmationButton } from '../../../utils/ui/confirmation-button';
import { EnrichedGraph } from '../types';
import { GraphViewer } from './view';

interface GraphWrapperArgs {
    graph: EnrichedGraph<Graph>;
    reload: () => void;
    update: (graphs: Graph[]) => void;
}

export const GraphWrapperWithRechart: React.FC<GraphWrapperArgs> = ({ graph, reload, update }) => {
    const logger = useLogger('GraphWrapperWithRechart');
    const navigate = useNavigate();
    const client = useApiClient();
    const deleteCb = useAsyncCallback((id: ApiUUID) => {
        return client.DELETE('/graphs/{id}', { params: { path: { id } } });
    });
    const moveGraph = useAsyncCallback(async (id: ApiUUID, direction: MoveDirection) => {
        const response = await client.PUT('/graphs/{id}/move', { params: { path: { id } }, body: { direction } });
        if (response.response.status === 200 && response.data) {
            update(response.data.results);
        }
    });
    return (
        <Box direction="column">
            <Heading level={3}>{graph.name}</Heading>
            <GraphViewer graph={graph} />
            <Box direction="row" justify="center">
                <Button
                    icon={<FormUp />}
                    onClick={() => {
                        catchAndLog(moveGraph.execute(graph.id, 'up'), 'Moving a graph up', logger);
                    }}
                />
                <Button
                    icon={<FormDown />}
                    onClick={() => {
                        catchAndLog(moveGraph.execute(graph.id, 'down'), 'Moving a graph down', logger);
                    }}
                />
                <Button
                    icon={<Edit />}
                    onClick={() => {
                        catchAndLog(Promise.resolve(navigate(`graph/${graph.id}`)), "Navigate to edit graph", logger);
                    }}
                />
                <ConfirmationButton
                    color="accent-4"
                    icon={<Trash />}
                    onConfirm={() => {
                        deleteCb.execute(graph.id).then(
                            () => {
                                reload();
                            },
                            (error: unknown) => {
                                logger.error('Error deleting graph', error);
                            },
                        );
                    }}
                />
            </Box>
        </Box>
    );
};
