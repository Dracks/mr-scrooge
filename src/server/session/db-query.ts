import { Op, WhereOptions } from 'sequelize';

export const queryOwnerId = (groupsId: number[]): WhereOptions<{ groupOwnerId: number }> => ({
    groupOwnerId: { [Op.in]: groupsId },
});
