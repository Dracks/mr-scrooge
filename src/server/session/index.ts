export { GqlGroupsId } from './decorators/gql-groups-id';
export { WrongOwnerId } from './gql-objects/wrong-owner-id.object';
export { AllowRoles, Role } from './guard/roles.decorator';
