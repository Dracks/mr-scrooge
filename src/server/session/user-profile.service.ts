import { Injectable, Logger } from "@nestjs/common";
import {InjectRepository} from '@mikro-orm/nestjs';
import {EntityRepository} from '@mikro-orm/core'


import { UserEntity } from "./entities/user.entity";
import { PasswordService } from "./password.service";


const SECRET_KEY = '2&_7krtq544)s^9=8i$om!u+9ump*)#-7rcrka9pjld2nnk6p='

const emptyProfile: Pick<UserEntity, 'isSuperuser' | 'lastName'| 'firstName'| 'email' | 'isStaff' | 'isActive'> = {
    isSuperuser: false, 
    lastName: '', 
    firstName: '', 
    email: '', 
    isStaff: false,
    isActive: false,
}

@Injectable()
export class UserProfileService{
    private readonly logger = new Logger(this.constructor.name)

    constructor(
        @InjectRepository(UserEntity) private readonly userModel: EntityRepository<UserEntity>, 
        readonly passwordService: PasswordService
       ){}

    public async getUserProfile(userId: number): Promise<UserEntity| undefined> {
        this.logger.log( {userId},'getUserProfile',)
        const user = await this.userModel.findOne({id: userId});

        return user ?? undefined;
    }

    public async existsUser(userId: number): Promise<boolean>
    {
        this.logger.log( {userId},'existsUser',)
        const exists = await this.userModel.findOne({id: userId});
        if (exists){
            return true;
        }
        return false;
    }

    public async validateUser(username: string, password: string): Promise<UserEntity | undefined> {
        this.logger.log( {username, password}, 'validateUser');
        const users = await this.userModel.find({$and: [{$or:[{username}, {email: username}]}, {isActive: true}]});

        const user = await users.find(async ({password: oldHash})=> {
            let isValid = false;
            if (oldHash){
                if (oldHash[0] !== '$'){
                    this.logger.log( {user}, 'password is a django one')

                    isValid = await this.passwordService.validateDjango(password, oldHash);
                } else {
                    isValid = await this.passwordService.validate(password, oldHash);
                }
            }
            this.logger.log( {user, isValid}, 'IS valid?');
            return isValid;
        })
        return user;
    }

    public async addUser(username: string, password: string, options: Partial<Omit<UserEntity, 'username' | 'password'>>={}): Promise<UserEntity> {
        this.logger.log( {username, password}, 'addUser');
        const hashedPassword = await this.passwordService.hash(password)
        const user = await this.userModel.create({username, password: hashedPassword, dateJoined: new Date(), ...emptyProfile, ...options,});
        this.logger.log({user}, 'Hey this is the user')
        await this.userModel.persistAndFlush(user);
        return user
    }
}