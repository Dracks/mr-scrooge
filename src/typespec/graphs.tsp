import "@typespec/http";
import "@typespec/rest";
import "@typespec/openapi3";

import "./global.tsp";

using TypeSpec.Http;
using TypeSpec.OpenAPI;
using TypeSpec.Rest;

namespace MrScrooge;

enum GraphKind {
  bar,
  line,
  pie,
}

enum GraphGroupType {
  day,
  labels,
  month,
  sign,
  year,
}

enum GraphDateRange {
  all,
  halfYear,
  oneMonth,
  oneYear,
  sixYears,
  twoYears,
}

model GraphParam {
  groupOwnerId: UUID;
  name: string;
  kind: GraphKind;
  labelFilterId?: UUID;
  dateRange: GraphDateRange;
  group: Group;
  horizontalGroup?: HorizontalGroup;
}

model Graph {
  id: UUID;
  ...GraphParam;
}

model Group {
  group: GraphGroupType;
  hideOthers: boolean | null;
  labels: UUID[] | null;
}

model HorizontalGroup {
  ...Group;
  accumulate: boolean | null;
}


model NotFoundLabels {
  ...NotFound;
  validLabels: UUID[];
  invalidLabels: UUID[];
}

@tag("graphs")
@route("/graphs")
interface ApiGraphs {
  @get list(
    @query graphIds?: [UUID],
    @query cursor?: string,
    @query limit?: int32 = 100,
  ): PageList<Graph>;
  @post new(...GraphParam): Graph | InvalidGroupOwnerId | NotFoundLabels;
  @route("/{id}") @patch update(@path id: UUID, ...GraphParam): Graph | InvalidGroupOwnerId | NotFoundLabels | NotFound;
  @route("/{id}") @delete delete(@path id: UUID): boolean;
}
