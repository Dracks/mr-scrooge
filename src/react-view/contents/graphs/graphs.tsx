import { Grid } from 'grommet';
import React from 'react';

import { useApiClient } from '../../api/client';
import { usePagination } from '../../api/pagination';
import { useLogger } from '../../utils/logger/logger.context';
import { LoadingPage } from '../../utils/ui/loading';
import { useLabelsContext } from '../common/label.context';
import { AddGraphPlaceholder } from './graph-with-rechart/add-graph-placeholder';
import { enrichGraph } from './graph-with-rechart/enrich-graph';
import { GraphWrapperWithRechart } from './graph-with-rechart/graph';

export const Graphs: React.FC = () => {
    const client = useApiClient();
    const logger = useLogger('Graphs');
    const graphs = usePagination(
        async next => {
            const { data } = await client.GET('/graphs', { params: { query: { cursor: next } } });
            if (data) {
                return data;
            } else {
                throw Error("Get graphs didn't had data");
            }
        },
        { autostart: true, hash: graph => graph.id },
    );
    const { labels } = useLabelsContext();

    const responseGraphList = graphs.loadedData;

    if (graphs.status == 'loading') {
        return <LoadingPage />;
    } else if (graphs.status === 'error') {
        logger.error('Error loading the graphs', graphs.error);
        return <div>Error loading the graphs </div>;
    } else {
        const enrichedGraphs = responseGraphList.map(graph => enrichGraph(graph, labels));
        enrichedGraphs.sort((a, b) => {
            return a.order - b.order;
        });
        return (
            <Grid columns={'450px'} gap="small">
                {enrichedGraphs.map(graph => (
                    <GraphWrapperWithRechart
                        key={graph.id}
                        graph={graph}
                        reload={() => {
                            graphs.reset();
                        }}
                        update={updatedGraphs => {
                            graphs.update(updatedGraphs);
                        }}
                    />
                ))}
                <AddGraphPlaceholder />
            </Grid>
        );
    }
};
