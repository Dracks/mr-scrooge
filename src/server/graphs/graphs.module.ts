import { Module } from '@nestjs/common';
import { SequelizeModule } from '@nestjs/sequelize';

import {
    GraphGroupLabelsModel,
    GraphGroupModel,
    GraphHorizontalGroupLabelsModel,
    GraphHorizontalGroupModel,
    GraphModel,
} from './models/graph.model';
import { LabelModel,LabelTransactionModel } from './models/label.model';
import { GraphResolver } from './resolvers/graph.resolver';
import { LabelResolver } from './resolvers/label.resolver';
import { GraphService } from './services/graph.service';
import { LabelService } from './services/label.service';

@Module({
    imports: [
        SequelizeModule.forFeature([
            LabelModel,
            LabelTransactionModel,
            GraphModel,
            GraphGroupModel,
            GraphHorizontalGroupModel,
            GraphGroupLabelsModel,
            GraphHorizontalGroupLabelsModel,
        ]),
    ],
    providers: [GraphService, LabelService, LabelResolver, GraphResolver],
    exports: [GraphService, LabelService],
})
export class GraphsModule {}
