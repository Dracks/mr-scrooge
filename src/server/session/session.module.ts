import { Module } from '@nestjs/common';
import { SequelizeModule } from '@nestjs/sequelize'

import { UserModel } from './models/user.model';
import { PasswordService } from './password.service';
import { SessionResolver } from './resolvers/session.resolver';
import { UserProfileService } from './user-profile.service';

@Module({
	imports: [SequelizeModule.forFeature([UserModel])],
	providers: [SessionResolver, UserProfileService, PasswordService],
	exports: [UserProfileService],
})
export class SessionModule { }