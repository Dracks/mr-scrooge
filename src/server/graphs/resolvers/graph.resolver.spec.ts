/* eslint-disable @typescript-eslint/no-unused-vars */
/* eslint-disable no-underscore-dangle */
import { FastifyAdapter,NestFastifyApplication } from '@nestjs/platform-fastify';
import { Test, TestingModule } from '@nestjs/testing';
import { FastifyInstance } from 'fastify';
import request from 'supertest-graphql';

import { TestDbModule } from '../../common/test-db.module';
import { GetGraphsByIdDocument, GetGraphsDocument, GQLGetGraphsByIdQuery, GQLGetGraphsByIdQueryVariables, GQLGetGraphsQuery, GQLGetGraphsQueryVariables,  GQLGraph, GQLGraphDateRange, GQLGraphGroup, GQLGraphKind,GQLNewGraphMutation, GQLNewGraphMutationVariables, GQLUpdateGraphMutation, GQLUpdateGraphMutationVariables, NewGraphDocument, UpdateGraphDocument } from '../../common/test-graphql/generated';
import { getGraphQLTestModule } from '../../common/test-graphql/graph-ql.module';
import {WrongOwnerId} from '../../session'
import { InvalidGraph } from '../gql-objects/invalid-graph.object';
import { GraphsModule } from '../graphs.module';
import { GraphDateRange, GraphGroup, GraphKind } from '../models/graph.model';
import { Graph, GraphService } from '../services/graph.service';
import { LabelService } from '../services/label.service';
import { GQLGraphFactory, GraphFactory } from '../test-data/graph.factory';
import { labelFactory } from '../test-data/label.factory';
import { GraphResolver } from './graph.resolver';

describe(GraphResolver.name, () => {
    let app!: NestFastifyApplication;
    let server!: FastifyInstance;
    let graphService!: GraphService;

    beforeEach(async () => {
        const moduleFixture: TestingModule = await Test.createTestingModule({
            imports: [TestDbModule, getGraphQLTestModule(() => ({
                groupsId: [1]
            })), GraphsModule],
        }).compile();

        app = moduleFixture.createNestApplication<NestFastifyApplication>(new FastifyAdapter());

        await app.init();
        await app.getHttpAdapter().getInstance().ready();
        server = app.getHttpAdapter().getHttpServer();
        graphService = app.get(GraphService)
    });

    afterEach(async () => {
        await app.close();
    });

    it('Starts the server', () => {
        expect(app).toBeTruthy();
    });

    describe('testing with some data', () => {
        beforeEach(async () => {
            await app.get(GraphService).createGraph(GraphFactory.build({name: 'first', id: 1}))
            await app.get(GraphService).createGraph(GraphFactory.build({
                dateRange: GraphDateRange.oneYear,
                group: {
                    group: GraphGroup.Labels,
                    hideOthers: false,
                    labels: [ 1, 2],
                },
                horizontalGroup: {
                    group: GraphGroup.Labels,
                    hideOthers: true,
                    labels: [ 3 , 4 ],
                },
                id: 34,
                kind: GraphKind.Line,
                name: 'huge graph',
            }))
            await app.get(GraphService).createGraph(GraphFactory.build({name: 'first', groupOwnerId: 3, id: 12344}))
        });

        it('Testing query the graphs', async () => {
            const response = await request<GQLGetGraphsQuery, GQLGetGraphsQueryVariables>(server).query(
                GetGraphsDocument,
            );
            expect(response.errors).toEqual(undefined);
            expect(response.data?.graphs).toHaveLength(2)
            expect(response.data?.graphs[0]).toMatchSnapshot();
            expect(response.data?.graphs[1]).toMatchSnapshot();
        });

        it('Testing query the graphs', async () => {
            const response = await request<GQLGetGraphsByIdQuery, GQLGetGraphsByIdQueryVariables>(server).query(
                GetGraphsByIdDocument,
                {graphsIds: [34]}
            );
            expect(response.errors).toEqual(undefined);
            expect(response.data?.graphs).toHaveLength(1)
            expect(response.data?.graphs[0].id).toEqual(34);
        });
    })

    describe ('Creating a new graph', ()=>{
        it('Testing the new graph', async ()=>{
            const {id: _id, ...newGraph} = GQLGraphFactory.build({
                kind: GQLGraphKind.Bar,
            })
            const response = await request<GQLNewGraphMutation, GQLNewGraphMutationVariables>(server).mutate(
                NewGraphDocument,
                {
                    graph: newGraph,
                }
            )
            expect(response.errors).toEqual(undefined);
            expect(response.data?.newGraph?.__typename).toEqual('Graph');
            const data = response.data?.newGraph as GQLGraph
            expect(data.id).toBeTruthy()
            expect(data).toMatchSnapshot()

            const dbGraph = await graphService.getGraphById(data.id)
            expect(dbGraph).toEqual({
                dateRange: "six",
                group: {
                    graphId: 1,
                    group: "month",
                    hideOthers: null,
                },
                groupOwnerId: 1,
                id: data.id,
                kind: "bar",
                labelFilter: null,
                name: "name: 0",
            })
        })

        it('Testing the new graph with invalid owner Id', async ()=>{
            // eslint-disable-next-line @typescript-eslint/no-unused-vars
            const {id: _id, ...newGraph} = GQLGraphFactory.build({ 
                groupOwnerId: 35,
                kind: GQLGraphKind.Bar,
            })
            const response = await request<GQLNewGraphMutation, GQLNewGraphMutationVariables>(server).mutate(
                NewGraphDocument,
                {
                    graph: newGraph,
                }
            )
            expect(response.errors).toEqual(undefined);
            expect(response.data?.newGraph?.__typename).toEqual('WrongOwnerId');
            const data = response.data?.newGraph as Partial<WrongOwnerId>
            expect(data.validOwners).toEqual([1])
        })
    });

    describe('UpdateGraph', ()=>{
        let existingGraphs!: Graph[] 
        let updatedGraph!: GQLGraph
        beforeEach(async ()=>{
            const graphsToCreate = [
                GraphFactory.build()
            ]
            existingGraphs = await Promise.all(graphsToCreate.map(graph => graphService.createGraph(graph))) 
            updatedGraph = GQLGraphFactory.build({
                dateRange: GQLGraphDateRange.HalfYear,
                group: {
                    group: GQLGraphGroup.Labels,
                    labels: [2,4]
                },
                horizontalGroup: {
                    group: GQLGraphGroup.Labels,
                    labels: [3,5]
                },
                id: existingGraphs[0].id,
                labelFilter: 1,
                name: 'Updated graph',
            })
        })
        it('testing updating a graph with invalid id', async ()=>{
            updatedGraph.id = -1;

            const response = await request<GQLUpdateGraphMutation, GQLUpdateGraphMutationVariables>(server).mutate(
                UpdateGraphDocument,
                { graph: updatedGraph }
            )

            expect(response.errors).toEqual(undefined)
            expect(response.data?.updateGraph.__typename).toEqual('InvalidGraph')
            const error = response.data?.updateGraph as unknown as InvalidGraph;
            expect(error.availableGraphsId).toEqual(existingGraphs.map(graph => graph.id))
        })

        it('testing updating a graph with invalid labels', async () => {
            const response = await request<GQLUpdateGraphMutation, GQLUpdateGraphMutationVariables>(server).mutate(
                UpdateGraphDocument,
                { graph: updatedGraph }
            )

            expect(response.errors).toEqual(undefined)
            expect(response.data?.updateGraph.__typename).toEqual('InvalidLabel')
        })

        it('testing with valid data', async ()=>{
            const labelsList = labelFactory.buildList(10)
            const labelService = app.get(LabelService)
            await Promise.all(labelsList.map(label => labelService.createLabel(label)))

            const response = await request<GQLUpdateGraphMutation, GQLUpdateGraphMutationVariables>(server).mutate(
                UpdateGraphDocument,
                { graph: updatedGraph }
            )

            expect(response.errors).toEqual(undefined)
            expect(response.data?.updateGraph.__typename).toEqual('Graph')
        })
    })
});
