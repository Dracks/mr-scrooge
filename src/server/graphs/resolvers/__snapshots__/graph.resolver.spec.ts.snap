// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphResolver Creating a new graph Testing the new graph 1`] = `
{
  "__typename": "Graph",
  "dateRange": "halfYear",
  "group": {
    "group": "Month",
    "hideOthers": null,
    "labels": null,
  },
  "groupOwnerId": 1,
  "horizontalGroup": null,
  "id": 1,
  "kind": "Bar",
  "labelFilter": null,
  "name": "name: 0",
}
`;

exports[`GraphResolver testing with some data Testing query the graphs 1`] = `
{
  "dateRange": "halfYear",
  "group": {
    "group": "Month",
    "hideOthers": null,
    "labels": null,
  },
  "groupOwnerId": 1,
  "horizontalGroup": null,
  "id": 1,
  "kind": "Pie",
  "labelFilter": null,
  "name": "first",
}
`;

exports[`GraphResolver testing with some data Testing query the graphs 2`] = `
{
  "dateRange": "oneYear",
  "group": {
    "group": "Labels",
    "hideOthers": false,
    "labels": [
      1,
      2,
    ],
  },
  "groupOwnerId": 1,
  "horizontalGroup": {
    "accumulate": false,
    "group": "Labels",
    "hideOthers": true,
    "labels": [
      3,
      4,
    ],
  },
  "id": 34,
  "kind": "Line",
  "labelFilter": null,
  "name": "huge graph",
}
`;
